### PR TITLE
feat(map_based_prediction): prediction with acc constraints

### PR DESCRIPTION
**Pull request by danielsanchezaran**
## Description

Implement a lateral acceleration-based pruning mechanism to eliminate unfeasible vehicle paths. See link for more details: [TIER IV INTERNAL LINK](https://tier4.atlassian.net/wiki/spaces/~7120207d8589d164d74a6d8162863204d3efe2/pages/2977566225/Map+based+prediction+with+lateral+acceleration+constraints)

Also, added the possibility of updating parameters in real time (only used for the parameters added by this PR). 

## Related links

[TIER IV INTERNAL LINK](https://tier4.atlassian.net/wiki/spaces/~7120207d8589d164d74a6d8162863204d3efe2/pages/2977566225/Map+based+prediction+with+lateral+acceleration+constraints)

## Tests performed

PSim with different vehicle speeds and Evaluator tests (I did not find degradation caused by this PR.
): [TIER IV INTERNAL LINK]( https://evaluation.tier4.jp/evaluation/reports/4b9c1ce4-30bf-5bbb-ae97-f559144ba701?project_id=prd_jt)
Example video:

The paths to turn right or left dissappear then the vehicle is close t\because it cannot realistically slow down on time to turn.

https://github.com/autowarefoundation/autoware.universe/assets/25967964/e66e806b-719e-4155-9c00-156438fc9c65

## Notes for reviewers

Requires these changes to launch: https://github.com/autowarefoundation/autoware_launch/pull/759

## Effects on system behavior

Eliminates predicted paths (for vehicles) that are unfeasible because they would generate large lateral accelerations

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: perception/map_based_prediction/README.md [hunk 1]
```diff
@@ -66,7 +66,7 @@ Lane change logics is illustrated in the figure below.An example of how to tune
 
 ### Tuning lane change detection logic
 
-Currently we provide two parameters to tune lane change detection:
+Currently we provide three parameters to tune lane change detection:
```
Comment (kyoichi-sugahara) [inline perception/map_based_prediction/README.md:69]: :heart:

File: perception/map_based_prediction/src/map_based_prediction_node.cpp [hunk 2]
```diff
@@ -972,16 +997,55 @@ void MapBasedPredictionNode::objectsCallback(const TrackedObjects::ConstSharedPt
...
         std::vector<PredictedPath> predicted_paths;
+        double min_avg_curvature = std::numeric_limits<double>::max();
+        PredictedPath path_with_smallest_avg_curvature;
+
         for (const auto & ref_path : ref_paths) {
           PredictedPath predicted_path = path_generator_->generatePathForOnLaneVehicle(
             yaw_fixed_transformed_object, ref_path.path);
           if (predicted_path.path.empty()) {
             continue;
           }
```
Comment (kyoichi-sugahara) [inline perception/map_based_prediction/src/map_based_prediction_node.cpp:1008]:
```suggestion
          if (predicted_path.path.empty()) continue;
```

File: perception/map_based_prediction/src/map_based_prediction_node.cpp [hunk 3]
```diff
@@ -972,16 +997,55 @@ void MapBasedPredictionNode::objectsCallback(const TrackedObjects::ConstSharedPt
...
-          predicted_path.confidence = ref_path.probability;
-          predicted_paths.push_back(predicted_path);
+
+          if (!check_lateral_acceleration_constraints_) {
```
Comment (kyoichi-sugahara) [inline perception/map_based_prediction/src/map_based_prediction_node.cpp:1010]: can you summarize if statement like this?
```
          if (!check_lateral_acceleration_constraints_ || lat_acceleration_satisfied) {
            predicted_path.confidence = ref_path.probability;
            predicted_paths.push_back(predicted_path);
```

File: perception/map_based_prediction/include/map_based_prediction/map_based_prediction_node.hpp [hunk 4]
```diff
@@ -237,6 +252,114 @@ class MapBasedPredictionNode : public rclcpp::Node
...
+        next_pose.position.y - current_pose.position.y);
+      arc_length += delta_s;
+
+      // Compute change in heading
+      tf2::Quaternion q_current, q_next;
+      tf2::convert(current_pose.orientation, q_current);
+      tf2::convert(next_pose.orientation, q_next);
+      double delta_theta = q_current.angleShortestPath(q_next);
+      // Handle wrap-around
+      if (delta_theta > 3.14159) {
```
Comment (kyoichi-sugahara) [inline perception/map_based_prediction/include/map_based_prediction/map_based_prediction_node.hpp:340]: how about using `M_PI` for 3.14159 ?

File: perception/map_based_prediction/include/map_based_prediction/map_based_prediction_node.hpp [hunk 5]
```diff
@@ -237,6 +252,114 @@ class MapBasedPredictionNode : public rclcpp::Node
...
+      double curvature = 0.0;
+      const auto p0 = getPoint(trajectory.at(i - std::min(idx_dist, i)));
+      const auto p1 = getPoint(trajectory.at(i));
+      const auto p2 = getPoint(trajectory.at(i + std::min(idx_dist, trajectory.size() - 1 - i)));
+      try {
+        curvature = calcCurvature(p0, p1, p2);
+      } catch (std::exception const & e) {
+        // ...code that handles the error...
+        RCLCPP_WARN(
+          rclcpp::get_logger("motion_velocity_smoother").get_child("trajectory_utils"), "%s",
```
Comment (kyoichi-sugahara) [inline perception/map_based_prediction/include/map_based_prediction/map_based_prediction_node.hpp:288]: need to fix here

File: perception/map_based_prediction/src/map_based_prediction_node.cpp (hunk 3 above)
Comment (kyoichi-sugahara) [reply perception/map_based_prediction/src/map_based_prediction_node.cpp:1010]: sorry please ignore this comment :bow:

File: perception/map_based_prediction/include/map_based_prediction/map_based_prediction_node.hpp (hunk 5 above)
Comment (danielsanchezaran) [reply perception/map_based_prediction/include/map_based_prediction/map_based_prediction_node.hpp]: Done, thanks!

File: perception/map_based_prediction/include/map_based_prediction/map_based_prediction_node.hpp (hunk 4 above)
Comment (danielsanchezaran) [reply perception/map_based_prediction/include/map_based_prediction/map_based_prediction_node.hpp]: Done!

File: perception/map_based_prediction/src/map_based_prediction_node.cpp (hunk 2 above)
Comment (danielsanchezaran) [reply perception/map_based_prediction/src/map_based_prediction_node.cpp]: Done

**Review by kyoichi-sugahara — APPROVED** [review]
LGTM!
